src/Workspaces/Core/Portable/EmbeddedLanguages/VirtualChars/VirtualCharSequence.cs

```diff
@@ -0,0 +1,147 @@
+    {
+        public static readonly VirtualCharSequence Empty = Create(ImmutableArray<VirtualChar>.Empty);
+
+        public static VirtualCharSequence Create(ImmutableArray<VirtualChar> virtualChars)
+            => new VirtualCharSequence(new ImmutableArrayChunk(virtualChars));
+
+        public static VirtualCharSequence Create(int firstVirtualCharPosition, string underlyingData, TextSpan underlyingDataSpan)
+            => new VirtualCharSequence(new StringChunk(firstVirtualCharPosition, underlyingData, underlyingDataSpan));
+
+        /// <summary>
+        /// The actual characters that this <see cref="VirtualCharSequence"/> is a portion of.
+        /// </summary>
+        private readonly Chunk _leafCharacters;
+
+        /// <summary>
+        /// The portion of <see cref="_leafCharacters"/> that is being exposed.
+        /// </summary>
+        private readonly TextSpan _span;
+
+        private VirtualCharSequence(Chunk sequence) : this(sequence, new TextSpan(0, sequence.Length))
+        {
+        }
+
+        private VirtualCharSequence(Chunk sequence, TextSpan span)
+        {
+            if (span.Start > sequence.Length)
+            {
+                throw new ArgumentException();
+            }
+
+            if (span.End > sequence.Length)
+            {
+                throw new ArgumentException();
+            }
+
+            _leafCharacters = sequence;
+            _span = span;
+        }
+
+        public int Length => _span.Length;
+        public VirtualChar this[int index] => _leafCharacters[_span.Start + index];
+
+        public bool IsDefault => _leafCharacters == null;
+        public bool IsEmpty => Length == 0;
+        public bool IsDefaultOrEmpty => IsDefault || IsEmpty;
+
+        public VirtualCharSequence GetSubSequence(TextSpan span)
+           => new VirtualCharSequence(
+               _leafCharacters, new TextSpan(_span.Start + span.Start, span.Length));
+
+        public VirtualChar First() => this[0];
+        public VirtualChar Last() => this[this.Length - 1];
+
+        public Enumerator GetEnumerator()
+            => new Enumerator(this);
+
+        public VirtualChar? FirstOrNullable(Func<VirtualChar, bool> predicate)
+        {
+            foreach (var ch in this)
+            {
+                if (predicate(ch))
+                {
+                    return ch;
+                }
+            }
+
+            return null;
+        }
+
+        public bool Contains(VirtualChar @char)
+            => IndexOf(@char) >= 0;
+
+        public int IndexOf(VirtualChar @char)
+        {
+            int index = 0;
+            foreach (var ch in this)
+            {
+                if (ch == @char)
+                {
+                    return index;
+                }
+
+                index++;
+            }
+
+            return -1;
+        }
+
+        public string CreateString()
+        {
+            var builder = PooledStringBuilder.GetInstance();
```

Comment:
>GetInstance [](start = 46, length = 11)

Could specify a size?

Reply:
nope.  not offered by PooledStringBuilder (since they're pooled)

Reply:
Never mind. Could add a `PooledStringBuilder.GetInstance(int)` as we have for `PooledArrayBuilder`, but that's nitpick.

---
In reply to: [262252886](https://github.com/dotnet/roslyn/pull/33834#discussion_r262252886) [](ancestors = 262252886)